File: src/lib/nodes/reporter.ts
import { ChatGoogleGenerativeAI } from "@langchain/google-genai";
import { AgentState } from "../types";

export const reporterNode = async (state: AgentState): Promise<Partial<AgentState>> => {
  const companyInfo = state.companyInfo || {};
  const scores = state.scores || {};
  const synth = state.synthesis || {};
  const comp = state.competitiveAnalysis || {};
  
  const resolvedName = companyInfo.name || state.companyInput || "the company";
  const verdict = state.verdict || "UNKNOWN";
  const confidence = state.confidence || 0;
  const weightedTotal = scores.weightedTotal || 5;
  const headline = state.headline || "No headline provided";
  const sector = companyInfo.sector || "Unknown sector";
  const country = companyInfo.country || "Unknown country";
  const description = companyInfo.description || "No description provided";
  const keyStrengths = synth.keyStrengths?.join(", ") || "None listed";
  const keyRisks = synth.keyRisks?.join(", ") || "None listed";
  
  const growth = scores.growth || 5;
  const moat = scores.moat || 5;
  const financialHealth = scores.financialHealth || 5;
  const sentiment = scores.sentiment || 5;
  const valuation = scores.valuation || 5;
  
  const investThesis = state.investThesis || "No thesis provided";
  const watchFor = state.watchFor?.join(", ") || "Nothing specified";
  const mainCompetitors = comp.mainCompetitors?.join(", ") || "None listed";
  const moatType = comp.moatType || "none";
  const moatStrength = comp.moatStrength || 5;

  const dateStr = new Date().toISOString().split('T')[0];

  try {
    const model = new ChatGoogleGenerativeAI({
      model: "gemini-2.5-flash",
      temperature: 0.3,
      maxOutputTokens: 1500,
    });

    const systemPrompt = "You are a senior equity research analyst writing a client-facing investment brief.\nWrite in clean markdown. Be direct and factual.\nEvery sentence must add information.";
    
    const userPrompt = `Write a full investment research brief for ${resolvedName} using this data:
  
  Verdict: ${verdict} | Score: ${weightedTotal}/10 | Confidence: ${confidence}%
  Headline: ${headline}
  Sector: ${sector} | Country: ${country}
  Description: ${description}
  Strengths: ${keyStrengths}
  Risks: ${keyRisks}
  Scores: Growth ${growth} | Moat ${moat} | Health ${financialHealth} | Sentiment ${sentiment} | Valuation ${valuation}
  Investment Thesis: ${investThesis}
  Watch For: ${watchFor}
  Competitors: ${mainCompetitors}
  Moat: ${moatType} (${moatStrength}/10)
  
  Use this EXACT markdown structure:
  
  # ${resolvedName} — Investment Research Brief
  ## Verdict: ${verdict} (${confidence}% confidence)
  > ${headline}
  ## Company Snapshot
  ## Investment Thesis
  ## Scorecard
  (markdown table with all 5 dimensions + weights)
  ## Key Strengths
  ## Key Risks  
  ## Competitive Position
  ## What to Watch
  ## Analyst Note
  ---
  *AI Investment Agent · ${dateStr} · Informational only*`;

    const response = await model.invoke([
      { role: "system", content: systemPrompt },
      { role: "user", content: userPrompt }
    ]);

    let content = typeof response.content === 'string' 
      ? response.content 
      : (Array.isArray(response.content) && response.content.length > 0 && 'text' in response.content[0]) 
        ? String((response.content[0] as any).text) 
        : '';
        
    // If Claude/Gemini wraps in code block, strip the backticks
    if (content.startsWith("\`\`\`markdown")) {
      content = content.replace(/^\`\`\`markdown\n?/, '').replace(/\`\`\`\n?$/, '').trim();
    } else if (content.startsWith("\`\`\`")) {
      content = content.replace(/^\`\`\`\n?/, '').replace(/\`\`\`\n?$/, '').trim();
    }
    
    return {
      report: content,
      streamLog: ["✅ Report generated"]
    };

  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    return {
      streamLog: [`❌ Error in Reporter Node: ${errorMessage}`],
      report: `# ${resolvedName} — Investment Research Brief\n\nFailed to generate report due to error: ${errorMessage}`
    };
  }
};

File: backend/nodes/reporter.py
"""
nodes/reporter.py — Investment Report Generator Node
Port of src/lib/nodes/reporter.ts
"""

import logging
from datetime import date
from langchain_groq import ChatGroq
from langchain_core.messages import SystemMessage, HumanMessage

from state import AgentState

logger = logging.getLogger(__name__)


async def reporter_node(state: AgentState) -> dict:
    company_info = state.get("companyInfo") or {}
    scores = state.get("scores") or {}
    synthesis = state.get("synthesis") or {}
    comp = state.get("competitiveAnalysis") or {}

    resolved_name = company_info.get("name") or state.get("companyInput") or "the company"
    verdict = state.get("verdict") or "UNKNOWN"
    confidence = state.get("confidence") or 0
    weighted_total = scores.get("weightedTotal") or 5
    headline = state.get("headline") or "No headline provided"
    sector = company_info.get("sector") or "Unknown sector"
    country = company_info.get("country") or "Unknown country"
    description = company_info.get("description") or "No description provided"

    key_strengths = ", ".join(synthesis.get("keyStrengths") or []) or "None listed"
    key_risks = ", ".join(synthesis.get("keyRisks") or []) or "None listed"

    growth = scores.get("growth") or 5
    moat = scores.get("moat") or 5
    financial_health = scores.get("financialHealth") or 5
    sentiment = scores.get("sentiment") or 5
    valuation = scores.get("valuation") or 5

    invest_thesis = state.get("investThesis") or "No thesis provided"
    watch_for = ", ".join(state.get("watchFor") or []) or "Nothing specified"
    main_competitors = ", ".join(comp.get("mainCompetitors") or []) or "None listed"
    moat_type = comp.get("moatType") or "none"
    moat_strength = comp.get("moatStrength") or 5

    date_str = date.today().isoformat()

    try:
        model = ChatGroq(
            model="openai/gpt-oss-120b",
            temperature=0.3,
            max_tokens=2048,
        )

        system_prompt = (
            "You are a senior equity research analyst writing a client-facing investment brief.\n"
            "Write in clean markdown. Be direct and factual.\n"
            "Every sentence must add information."
        )

        user_prompt = f"""Write a full investment research brief for {resolved_name} using this data:

Verdict: {verdict} | Score: {weighted_total}/10 | Confidence: {confidence}%
Headline: {headline}
Sector: {sector} | Country: {country}
Description: {description}
Strengths: {key_strengths}
Risks: {key_risks}
Scores: Growth {growth} | Moat {moat} | Health {financial_health} | Sentiment {sentiment} | Valuation {valuation}
Investment Thesis: {invest_thesis}
Watch For: {watch_for}
Competitors: {main_competitors}
Moat: {moat_type} ({moat_strength}/10)

Use this EXACT markdown structure:

# {resolved_name} — Investment Research Brief
## Verdict: {verdict} ({confidence}% confidence)
> {headline}
## Company Snapshot
## Investment Thesis
## Scorecard
(markdown table with all 5 dimensions + weights)
## Key Strengths
## Key Risks
## Competitive Position
## What to Watch
## Analyst Note
---
*AI Investment Agent · {date_str} · Informational only*"""

        response = await model.ainvoke([
            SystemMessage(content=system_prompt),
            HumanMessage(content=user_prompt),
        ])

        content = response.content if isinstance(response.content, str) else str(response.content)

        # Strip markdown code-block wrapper if the model wraps the whole report
        if content.startswith("```markdown"):
            content = content.removeprefix("```markdown").removesuffix("```").strip()
        elif content.startswith("```"):
            content = content.removeprefix("```").removesuffix("```").strip()

        return {
            "report": content,
            "streamLog": ["◈ BRIEF COMPILED"],
        }

    except Exception as exc:
        error_message = str(exc)
        return {
            "streamLog": [f"⊗ REPORTER FAULT — {error_message}"],
            "report": f"# {resolved_name} — Investment Research Brief\n\nFailed to generate report due to error: {error_message}",
        }
